File: 141.linked-list-cycle.go
package main

type ListNode struct {
	Next *ListNode
	Val  int
}

func fastSlowPointer(head *ListNode) bool {
	fast, slow := head, head

	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next

		if fast == slow {
			return true
		}
	}
	return false
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func hasCycle(head *ListNode) bool {
	const MAX int = 1000000

	for head != nil {
		if head.Val != MAX {
			head.Val = MAX
			head = head.Next
		} else {
			return true
		}
	}
	return false
}

// @lc code=end

File: 141.linked-list-cycle.rb
#
# @lc app=leetcode id=141 lang=ruby
#
# [141] Linked List Cycle
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end
def _hasCycle(head)
  until head.nil?
    return true if head.val.nil?
    head.val = nil
    head = head.next
  end
  false
end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
  fast, slow = head, head

  until fast.nil? || fast.next.nil?
    fast = fast.next.next
    slow = slow.next

    return true if fast == slow

  end

  false
end
# @lc code=end
